Mpuncel/sqlite pod process reporter

mpuncel opened the pull request.

anthonybishopric commented on pkg/kp/statusstore/podstatus/status.go:
```diff
@@ -5,53 +5,39 @@ import (
+// information about the last exit is exposed.
+type ProcessStatus struct {
+	LaunchableID launch.LaunchableID `json:"launchable_id"`
+	EntryPoint   string              `json:"name"`
```
Did you mean to change the JSON tag too?

mpuncel replied:
yes 🙏

anthonybishopric commented on pkg/preparer/podprocess/finish_exec.go:
```diff
@@ -0,0 +1,13 @@
+
+func (r *Reporter) FinishExec() []string {
+	return []string{
+		"/usr/bin/timeout",
```
This is not available on all systems. Can it be made configurable / opt-in?

mpuncel replied:
will make it configurable

anthonybishopric commented on pkg/preparer/podprocess/finish_exec.go:
```diff
@@ -0,0 +1,13 @@
+package podprocess
```
Do you envision this file growing much?

mpuncel replied:
I don't think so, are you suggesting I merge it into `reporter.go`? The reason I had it separate is because the preparer calls this function directly to determine the finish exec script to write, so it seemed separate from `reporter.go` in terms of purpose

anthonybishopric commented on pkg/preparer/podprocess/finish_output.go:
```diff
@@ -0,0 +1,48 @@
+	ExitStatus int `json:"exit_status"`
+}
+
+func (f FinishOutput) Insert(db *sql.DB) error {
```
SQL DB style recommendation - rather than putting this function on the data type, put it on a service type that has a function `InsertFinish(finishOutput, db)`. Constraining DB operations to a single interface will make it easier to stub later. 

It also results in the various queries / statements and sqlite implementation being consolidated to a single location.

anthonybishopric commented on pkg/preparer/podprocess/reporter.go:
```diff
@@ -0,0 +1,313 @@
+		}
+	}()
+	for rows.Next() {
+		err = rows.Scan(&id, &date, &podID, &podUniqueKey, &launchableID, &entryPoint, &exitCode, &exitStatus)
```
Move the query + scan to a `readFinishes` helper that returns a `[]finishRow` with these fields.

mpuncel replied:
I'm curious for the motivation here? the function doesn't do much else (the only other thing is `r.getLastID()`) and I don't anticipate other places to need to do the scan

mpuncel replied:
actually the submission is separate too, i'll move it to another function

anthonybishopric commented on pkg/preparer/podprocess/reporter.go:
```diff
@@ -0,0 +1,313 @@
+)
+
+const (
+	workspaceFileName     = "last_sql_id"
```
`"last_synced_finish_id"`

anthonybishopric commented on pkg/preparer/podprocess/reporter.go:
```diff
@@ -0,0 +1,313 @@
+		return nil, util.Errorf("Could not open sqlite database: %s", err)
+	}
+
+	sqlStmt := `
```
is it worth having a lightweight schema version number here? A follow up migration as trivial as adding a column will make this complicated real fast

anthonybishopric commented on pkg/preparer/podprocess/reporter.go:
```diff
@@ -0,0 +1,313 @@
+
+	sqlStmt := `
+	create table if not exists finishes (
+	    id integer not null primary key,
```
do you need an `autoincrement` here too?

mpuncel replied:
no: https://sqlite.org/faq.html#q1

mpuncel replied:
although maybe i'll add it for clarity assuming it works

mpuncel replied:
oh man i think I do need it for when we delete values: ```The new key will be unique over all keys currently in the table, but it might overlap with keys that have been previously deleted from the table. To create keys that are unique over the lifetime of the table, add the AUTOINCREMENT keyword to the INTEGER PRIMARY KEY declaration. ```

mpuncel commented:
btw I plan on adding an integration test for this once the env extractor PR is also merged, or perhaps I can add it to that one once this one is merged

anthonybishopric commented:
Nice changes, I think this is good to go 👍

mpuncel commented:
squashed the fixup commit, and fixed a bug in the migration code (have to insert a row when you create the table or the update won't do anything).

will merge on green

mpuncel commented:
fixed some errcheck complaints
